x-pack/plugins/cases/server/client/attachments/add_file.ts

```diff
@@ -0,0 +1,114 @@
+ * 2.0.
+ */
+
+import { SavedObjectsUtils } from '@kbn/core/server';
+
+import type { Readable } from 'stream';
+import type { Owner } from '../../../common/constants/types';
+import type { Case } from '../../../common/types/domain';
+import type { CasesClient, CasesClientArgs } from '..';
+import type { AddFileArgs } from './types';
+
+import { PostFileAttachmentRequestRt } from '../../../common/types/api';
+import { CaseCommentModel } from '../../common/models';
+import { createCaseError } from '../../common/error';
+import { validateMaxUserActions } from '../../common/validators';
+import { constructFileKindIdByOwner } from '../../../common/files';
+import { Operations } from '../../authorization';
+import { validateRegisteredAttachments } from './validators';
+import { decodeWithExcessOrThrow } from '../../common/runtime_types';
+import { buildAttachmentRequestFromFileJSON } from '../utils';
+
+/**
+ * Create a file attachment to a case.
+ *
+ * @ignore
+ */
+export const addFile = async (
+  addFileArgs: AddFileArgs,
+  clientArgs: CasesClientArgs,
+  casesClient: CasesClient
+): Promise<Case> => {
+  const { caseId, fileRequest } = addFileArgs;
+  const {
+    logger,
+    authorization,
+    persistableStateAttachmentTypeRegistry,
+    externalReferenceAttachmentTypeRegistry,
+    services: { userActionService },
+    fileService,
+  } = clientArgs;
+
+  let createdFile;
+
+  try {
+    const {
+      file,
+      filename,
+      mimeType: mime,
+    } = decodeWithExcessOrThrow(PostFileAttachmentRequestRt)(fileRequest);
+
+    // This will perform an authorization check to ensure the user has access to the parent case
+    const theCase = await casesClient.cases.get({
+      id: caseId,
+      includeComments: false,
+    });
+
+    const owner = theCase.owner;
+
+    await validateMaxUserActions({ caseId, userActionService, userActionsToAdd: 1 });
+
+    const savedObjectID = SavedObjectsUtils.generateId();
+
+    await authorization.ensureAuthorized({
+      operation: Operations.createComment,
+      entities: [{ owner, id: savedObjectID }],
+    });
+
+    createdFile = await fileService.create({
+      name: filename,
+      mime,
+      fileKind: constructFileKindIdByOwner(owner as Owner),
+      meta: { caseIds: [caseId], owner: [owner] },
+    });
+
+    await createdFile.uploadContent(file as Readable);
+
+    const commentReq = buildAttachmentRequestFromFileJSON({
+      owner,
+      fileMetadata: createdFile.toJSON(),
+    });
+
+    validateRegisteredAttachments({
+      query: commentReq,
+      persistableStateAttachmentTypeRegistry,
+      externalReferenceAttachmentTypeRegistry,
+    });
+
+    const createdDate = new Date().toISOString();
+
+    const model = await CaseCommentModel.create(caseId, clientArgs);
+
+    const updatedModel = await model.createComment({
+      createdDate,
+      commentReq,
+      id: savedObjectID,
+    });
+
+    return await updatedModel.encodeWithComments();
+  } catch (error) {
```

Comment:
Should we handle some file errors? They are declared here `src/plugins/files/server/file/errors.ts`. I think for our use case the `AbortedUploadError` should be enough. Example:

```
if (error instanceof AbortedUploadError) {
     throw new Boom.Boom(error, {
          statusCode: 499,
          message: error.message,
        });
}
```